tutorials/transfer_learning/README.md

```diff
@@ -0,0 +1,254 @@
+
+Setting `parallel` to `true` is an important performance enhancement as the GPU would otherwise spent significant time on idle waiting for the CPU data loading to complete. 
+
+## Fine-tune | 🐢 mode
+
+Load a pre-trained model: 
+
+```julia
+m = Metalhead.ResNet(18, pretrain = true).layers
+```
+
+Substitute the latest layers with ones adapted to the fine-tuning task:
+
+```julia
+m_tot = Chain(m[1], AdaptiveMeanPool((1, 1)), Flux.flatten, Dense(512 => 3)) |> device
+```
+
+Define an accuracy evaluation function:
+
+```julia
+function eval_f(m, deval)
+    good = 0
+    count = 0
+    for (x, y) in deval
+        good += sum(Flux.onecold(m(x)) .== Flux.onecold(y))
+        count += size(y, 2)
+    end
+    acc = round(good / count, digits = 4)
+    return acc
+end
+```
+
+Define a training loop for 1 epoch: 
+
+```julia
+function train_epoch!(m; ps, opt, dtrain)
+    for (x, y) in dtrain
+        grads = gradient(ps) do
+            Flux.Losses.logitcrossentropy(m(x), y)
+        end
+        update!(opt, ps, grads)
+    end
+end
+```
+
+Set learnable parameters and optimiser:
+
+```julia
+ps = Flux.params(m_tot[2:end]);
+opt = Adam(3e-4)
+```
+
+Train for a few epochs:
+
+```julia
+for iter = 1:8
+    @time train_epoch!(m_tot; ps, opt, dtrain)
+    metric_train = eval_f(m_tot, dtrain)
+    metric_eval = eval_f(m_tot, deval)
+    @info "train" metric = metric_train
+    @info "eval" metric = metric_eval
+end
+ 12.932525 seconds (3.01 M allocations: 7.001 GiB, 16.11% gc time)
+┌ Info: train
+└   metric = 0.9481
+┌ Info: eval
+└   metric = 0.9467
+```
+
+## Fine-tune | 🐇 mode
+
+In the previous fine-tuning, despite having only specified the last `Dense` layer as trainable parameters, we nonetheless ended computing the gradients over the entire model. 
+
+To avoid these unnecessary computations, we can split our model in two: 
+- The original pre-trained core, for which we don't want to compute gradients
+- The new final layers, for which gradients are needed. 
+
+```julia
+m_infer = deepcopy(m[1]) |> device
+m_tune = Chain(AdaptiveMeanPool((1, 1)), Flux.flatten, Dense(512 => 3)) |> device
+```
+
+Only minimal adaptations are then needed to the eval and training functions: 
+
+```julia
+function eval_f(m_infer, m_tune, deval)
+    good = 0
+    count = 0
+    for (x, y) in deval
+        good += sum(Flux.onecold(m_tune(m_infer(x))) .== Flux.onecold(y))
+        count += size(y, 2)
+    end
+    acc = round(good / count, digits = 4)
+    return acc
+end
+```
+
+```julia
+function train_epoch!(m_infer, m_tune; ps, opt, dtrain)
+    for (x, y) in dtrain
+        infer = m_infer(x)
+        grads = gradient(ps) do
+            Flux.Losses.logitcrossentropy(m_tune(infer), y)
+        end
+        update!(opt, ps, grads)
+    end
+end
+```
+
+```julia
+ps = Flux.params(m_tune);
+opt = Adam(3e-4)
+```
```

Comment:
let's use the new Optmisers.jl interface instead of the params-based one

Reply:
Metalhead compat forces use of Flux v0.13.4, for which my understanding was that the new Optimisers.jl wasn't yet in place. 
There were some pending issues on weights import with Metalhead which once fixed, should allow to bump Flux compat, and I'd migrate the tutorial to Optimisers/explicit gradients once done.

Reply:
A new Metalhed version just got released

Reply:
Updating to latest Flux brought some new challenges with MLUtils! 

Creating the DataLoaders with `parallel=true` results in stalled / hanging julia session, at least on Windows: https://github.com/jeremiedb/model-zoo/blob/51e6a091aa9d790d3162cfe8a106ad45aa0beac1/tutorials/transfer_learning/transfer_learning.jl#L35-L56
This occurs when running manually from the REPL, but works fine when launched as a script. Looks like it might be related to https://github.com/JuliaML/MLUtils.jl/issues/142

Another "annoyance" is the need to have a `collect` in the `getobs` recipe, as mentionned in https://github.com/JuliaML/MLUtils.jl/issues/139, as it seems opposite the the typical Julia pattern where explicit `collect` isn't typically needed.

Would you be fine moving forward with the new Flux/MLUtils/explicit gradient approach, considering the above few caveats? I'd be fine going forward with latest versions and get these few gotchas ironed out.

Reply:
PR has been updated to explicit gradients

Reply:
@CarloLucibello Do you feel this explicit gradient refresh is robust enough to go forward, or should first figure out https://github.com/JuliaML/MLUtils.jl/issues/148 and others cavats?